src/client/pydaos/pydaos_shim.c

```diff
@@ -445,6 +445,268 @@ __shim_handle__cont_close(PyObject *self, PyObject *args)
 	return PyInt_FromLong(rc);
 }
 
+#define ITER_NR		96
+
+static int
+oit_mark(daos_handle_t oh, daos_handle_t oit)
+{
+	char			*buf = NULL;
+	daos_anchor_t		anchor = {0};
+	daos_key_desc_t		kds[ITER_NR];
+	d_sg_list_t		sgl;
+	d_iov_t			sg_iov;
+	daos_size_t		buf_size = ITER_NR * 256;
+	struct pydaos_df	entry;
+	size_t			size = sizeof(entry);
+	d_iov_t			marker;
+	bool			mark_data = true;
+	int			rc;
+
+	D_ALLOC(buf, buf_size);
+	if (buf == NULL)
+		goto out;
+
+	d_iov_set(&marker, &mark_data, sizeof(mark_data));
+	d_iov_set(&sg_iov, buf, buf_size);
+	sgl.sg_nr		= 1;
+	sgl.sg_nr_out		= 0;
+	sgl.sg_iovs		= &sg_iov;
+
+	while (!daos_anchor_is_eof(&anchor)) {
+		uint32_t	nr = ITER_NR, i;
+		void		*ptr;
+
+		memset(buf, 0, buf_size);
+		rc = daos_kv_list(oh, DAOS_TX_NONE, &nr, kds, &sgl, &anchor, NULL);
+		if (rc)
+			goto out;
+		if (nr == 0)
+			continue;
+
+		for (ptr = buf, i = 0; i < nr; i++) {
+			char *key;
+
+			D_ALLOC(key, kds[i].kd_key_len + 1);
+			if (key == NULL)
+				return -DER_NOMEM;
```

Comment:
We need to free `buf` that was alloc'ed above. Maybe?
```suggestion
			D_ALLOC(key, kds[i].kd_key_len + 1);
			if (key == NULL) {
				rc = -DER_NOMEM;
				goto out;
			}
```